sys/kern/dev_procstat.c

```diff
@@ -0,0 +1,175 @@
+#include <sys/devfs.h>
+#include <sys/vnode.h>
+#include <sys/linker_set.h>
+#include <sys/queue.h>
+#include <sys/proc.h>
+#include <sys/thread.h>
+#include <sys/malloc.h>
+#include <stdio.h>
+
+/* Implementation of /dev/procstat
+ *
+ * When /dev/procstat is opened it takes a snapshot of info of all existing
+ * processes.
+ *
+ * During read calls to /dev/procstat this info can be read.
+ *
+ * Example:
+ * euid   pid    ppid    pgrp   session  state    td_name  elfpath
+ * 0       1       0       1       1       R       init    /bin/ksh
+ */
+
+/* maximum size of output string of proc info */
+#define MAX_P_STRING (TD_NAME_MAX + PATH_MAX + 5 * 10 + 1 + 10)
+
+/* maximum amount of processes that procstat can handle */
+#define MAX_PROC 40
+
+static char state[4] = {
+  [PS_NORMAL] = 'R',
+  [PS_STOPPED] = 'S',
+  [PS_DYING] = 'D',
+  [PS_ZOMBIE] = 'Z',
+};
+
+typedef struct proc_info {
+  pid_t pid;
+  pid_t ppid;
+  uid_t uid;
+  pgid_t pgrp;
+  sid_t sid;
+  proc_state_t state;
+  char td_name[TD_NAME_MAX];
+  char elfpath[PATH_MAX];
```

Comment:
What about allocating the `elfpath` dynamically (using for example `kstrndup`)?

Reply:
I was a bit scared about doing it dynamically because we hold lock here. I changed that and it works. And I think it should because proc lock does nothing with malloc. Could you confirm that?

Reply:
It's safe to do that.